grunttasks/xssmitigate.js

```diff
@@ -0,0 +1,66 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+var xss = require('xss');
+
+module.exports = function (grunt) {
+  grunt.registerTask('xss-flag', 'Flag potentially harmful files', [
+    'clean',
+    'selectconfig:dist',
+    'l10n-generate-pages',
+    'l10n-json-to-html',
+    'xss-parse'
+    ]
+  );
+
+  grunt.registerMultiTask('xss-parse', 'Parse HTML files to check for XSS vectors', function () {
+    var templateRegex = /\%\(\w+\)s/;
+    this.files.forEach(function (file) {
+      // iterate through each HTML file in the .tmp folder
+      var src = file.src[0];
+      var contents = grunt.file.read(src);
+      // whitelist of allowed tags and attributes, nothing else is allowed
+      var options = {
```

Comment:
@TDA: Ah right, sorry, I missed that `src` was used inside the loop. Ignore me.

For the errors, I was thinking maybe something like this that loops over all files and reports all errors and fails at the end if there was 1+ error found:

```js
  grunt.registerMultiTask('xss-parse', 'Parse HTML files to check for XSS vectors', function () {
    var templateRegex = /\%\(\w+\)s/;
    var hasErrors = 0;
    this.files.forEach(function (file) {
      // iterate through each HTML file in the .tmp folder
      var src = file.src[0];
      var contents = grunt.file.read(src);
      // whitelist of allowed tags and attributes, nothing else is allowed
      var options = {
        whiteList: {
          a: ['href', 'id'],
          em: [],
          span: ['id', 'tabindex'],
          strong: []
        },
        onTagAttr: function (tag, name, value, isWhiteAttr) {
          // On encountering any tag, check if the tag has a valid attribute.
          // If the attribute is valid, check for its value, flag if illegal
          // Illegal includes any mention of the string 'javascript', case insensitive
          // Templated strings are allowed i.e strings of the form '%(termsURI)s'
          // If these are considered to be not secure, we should change
          if (isWhiteAttr && xss.safeAttrValue(tag, name, value) === '' && ! templateRegex.test(value)) {
            grunt.log.writeln('%s: INVALID VALUE FOR ATTRIBUTE <%s %s="%s">', src, tag, name, value);
            hasErrors += 1;
          }
        },
        onIgnoreTag: function (tag, html, options) {
          // On encountering any illegal tag, flag it
          grunt.log.writeln('%s: INVALID TAGS found <%s>', tag);
          hasErrors += 1;

        },
        onIgnoreTagAttr: function (tag, name, value, isWhiteAttr) {
          // On encountering any illegal attribute, flag it
          grunt.log.writeln('%s: INVALID TAG ATTRIBUTES FOUND <%s %s="">', src, tag, name);
          hasErrors += 1;
        }
      };
      xss(contents, options);
    });
    if (hasErrors !== 0) {
      grunt.fail.fatal('Found ' + hasErrors + ' ' + grunt.util.pluralize(hasErrors, 'error/errors') + ' in '+ this.files.length + ' files');
    }
  });
```

#### Output:

```
✗ grunt xss-parse
Running "xss-parse:dist" (xss-parse) task
.tmp/i18n/fy/client.html: INVALID TAG ATTRIBUTES FOUND <a title="">
.tmp/i18n/fy/client.html: INVALID VALUE FOR ATTRIBUTE <a href="javascript:alert(1)">
.tmp/i18n/fy/client.html: INVALID TAG ATTRIBUTES FOUND <a onclick="">
Fatal error: Found 3 errors in 104 files


Execution Time (2015-07-14 17:24:57 UTC)
loading tasks   1.6s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 96%
xss-parse:dist  56ms  ▇▇▇▇ 3%
Total 1.6s
```